### ISSUE TITLE
Improve error handling of crashing rules

### ISSUE THREAD

**Issue by AriPerkkio** [eslint/eslint#13872]
**The version of ESLint you are using.**

`ESLint: 7.12.1`

**The problem you want to solve.**

Error handling of crashing rules could be improved. By a "crashing rule" I mean a case where ESLint rule throws an error instead of using the `context.report` API. These issues are often caused by unsafe accessing of a node property, e.g. `node.callee.name` -> `Uncaught TypeError: Cannot read property 'name' of undefined`. I've noticed that these are quite common in various ESlint plugins.

Current functionality when erroneous rule causes linter to crash:
1. Cause of the crash is not clear. Users have to study the stack trace in order to figure out which rule caused the crash.
2. Every other valid linting problems are lost. If rule A is reporting valid linting problem, it is not visible due to rule B causing linter crash.

Let's use real-world example from `eslint-plugin-react-hooks`, [Bug: eslint-plugin-react-hooks: Cannot read property 'type' of undefined at analyzePropertyChain #20204](https://github.com/facebook/react/issues/20204).

<details>
    <summary>Current behavior</summary>

ESLint error message does not indicate which rule caused the crash.

```bash
$ node ./node_modules/.bin/eslint ./lib


Oops! Something went wrong! :(

ESLint: 7.12.1

TypeError: Cannot read property 'type' of undefined
Occurred while linting /<root-path-removed>/eslint-remote-tester/lib/ui/hooks/useOnExit.ts:5
    at analyzePropertyChain (/<root-path-removed>/eslint-remote-tester/node_modules/eslint-plugin-react-hooks/cjs/eslint-plugin-react-hooks.development.js:2235:12)
    at analyzePropertyChain (/<root-path-removed>/eslint-remote-tester/node_modules/eslint-plugin-react-hooks/cjs/eslint-plugin-react-hooks.development.js:2264:20)
    at /<root-path-removed>/eslint-remote-tester/node_modules/eslint-plugin-react-hooks/cjs/eslint-plugin-react-hooks.development.js:1297:34
    at Array.forEach (<anonymous>)
    at visitFunctionWithDependencies (/<root-path-removed>/eslint-remote-tester/node_modules/eslint-plugin-react-hooks/cjs/eslint-plugin-react-hooks.development.js:1277:43)
    at visitCallExpression (/<root-path-removed>/eslint-remote-tester/node_modules/eslint-plugin-react-hooks/cjs/eslint-plugin-react-hooks.development.js:1737:11)
    at /<root-path-removed>/eslint-remote-tester/node_modules/eslint/lib/linter/safe-emitter.js:49:25
    at Array.forEach (<anonymous>)
    at Object.emit (/<root-path-removed>/eslint-remote-tester/node_modules/eslint/lib/linter/safe-emitter.js:47:38)
    at NodeEventGenerator.applySelector (/<root-path-removed>/eslint-remote-tester/node_modules/eslint/lib/linter/node-event-generator.js:254:26)
```

Every other valid linting error is lost:

![eslint-crash-current-behavior](https://user-images.githubusercontent.com/14806298/99876710-1858e300-2c01-11eb-9e96-2d3b40bd3c7a.png)

</details>

By fixing the first issue users would know what rule caused the issue. After figuring out how ESlint runs the rules I created a POC for this fix.

<details>
    <summary>Crashing rule included in the error message</summary>

```bash
$ node ./node_modules/.bin/eslint ./lib


Oops! Something went wrong! :(

ESLint: 7.12.1

TypeError: Cannot read property 'type' of undefined caused by rule "react-hooks/exhaustive-deps"
Occurred while linting /<removed-root-path>/eslint-remote-tester/lib/ui/hooks/useOnExit.ts:5
    at /<removed-root-path>/eslint-remote-tester/node_modules/eslint/lib/linter/safe-emitter.js:54:31
    at Array.forEach (<anonymous>)
    at Object.emit (/<removed-root-path>/eslint-remote-tester/node_modules/eslint/lib/linter/safe-emitter.js:47:38)
    at NodeEventGenerator.applySelector (/<removed-root-path>/eslint-remote-tester/node_modules/eslint/lib/linter/node-event-generator.js:254:26)
    at NodeEventGenerator.applySelectors (/<removed-root-path>/eslint-remote-tester/node_modules/eslint/lib/linter/node-event-generator.js:283:22)
    at NodeEventGenerator.enterNode (/<removed-root-path>/eslint-remote-tester/node_modules/eslint/lib/linter/node-event-generator.js:297:14)
    at CodePathAnalyzer.enterNode (/<removed-root-path>/eslint-remote-tester/node_modules/eslint/lib/linter/code-path-analysis/code-path-analyzer.js:711:23)
    at /<removed-root-path>/eslint-remote-tester/node_modules/eslint/lib/linter/linter.js:966:32
    at Array.forEach (<anonymous>)
    at runRules (/<removed-root-path>/eslint-remote-tester/node_modules/eslint/lib/linter/linter.js:961:15)
```

However every other valid linting error is still lost:
![eslint-crash-rule-included](https://user-images.githubusercontent.com/14806298/99876719-2b6bb300-2c01-11eb-9815-5cdc4373a519.png)

</details>

By fixing the second issue users would still be able to see valid linting errors. Instead of throwing an Error from `linter.js` it could parse `safe-emitter.js`'s error and report it via `context.report`. I'm not sure whether this is against principles of ESlint but this would surely improve user experience.

<details>
    <summary>Crashing rule reported via `context.report`</summary>

These examples also contain fix for the first issue.

Other linting errors are not lost:
```bash
$ node ./node_modules/.bin/eslint ./lib

/<removed-rooth-path>/eslint-remote-tester/lib/ui/hooks/useOnExit.ts
  5:5   warning  Cannot read property 'type' of undefined caused by rule "react-hooks/exhaustive-deps"  react-hooks/exhaustive-deps
  8:11  warning  'id' is assigned a value but never used                                                @typescript-eslint/no-unused-vars

✖ 2 problems (0 errors, 2 warnings)
```
![eslint-crash-reported](https://user-images.githubusercontent.com/14806298/99876726-345c8480-2c01-11eb-8329-d8eb8b5166bd.png)

</details>

**Your take on the correct solution to problem.**

I would like to see the first issue (include rule name in error message) fixed. The second is not that clear and requires discussion.

**Are you willing to submit a pull request to implement this change?**
Yes! The POCs above are working OK in my local environment. Before I go any further with the implementation I would like to hear opinnions from the maintainers. The main goal here is to improve user and development experience of ESlint and its community plugins.

**Comment by anikethsaha** [comment]
Thanks for the issue. 

Few questions

- this will be a rule so I suppose when user adds the rule , those errors should be handled by the rule else it should be handled like it used to, right ?
- What about the stack trace ? adding the whole stack trace in rule message is not a good idea I think.

**Comment by AriPerkkio** [comment]
Thanks for the questions @anikethsaha, happy to answer them!

> * This will be implemented as a rule right ?

This would not be a rule. Instead these changes would be made to ESLint core. The main problem here is how ESLint is running third party rules without proper error handling. Or at least the error handling could be improved. 

Currently when a plugin contains rule which crashes it will throw an error here:
https://github.com/eslint/eslint/blob/110cf962d05625a8a1bf7b5f4ec2194db150eb32/lib/linter/safe-emitter.js#L45

Since these possible errors are not caught, whole linting process will crash and errors are printed as shown in the first post. Even a single crashing rule leads to a situation where all valid linting results are lost, due to linter crash. We could improve this so that a crashing rule would not crash whole linting process. 

> * What about the stack trace ? adding the whole stack trace in rule message is not a good idea I think.

In the POCs I've only included the `error.message` (e.g. `Cannot read property 'type' of undefined`) and the rule ID. The rule id could be passed to `safe-emitter.js` here:
https://github.com/eslint/eslint/blob/110cf962d05625a8a1bf7b5f4ec2194db150eb32/lib/linter/linter.js#L933-L941
At that point we have rule ID and "rule method" in the same scope. Passing it as new argument to `emitter.on` required small changes. 

So as summary the current issues are:
1. A crashing rule results into error message which does not indicate the crashing rule
2. A single rule can crash the whole linting process

The first issue could be fixed easily and doesn't change ESLint core that much. It would still throw the error and crash whole linting process as before. Now the error message would indicate the erroneous rule clearly.

The second issue requires quite major changes to ESLint core - not much code though. I'm quite unsure if reporting crashing rules with `context.report` API is ideal. How could these errors be handled and reported without throwing the error from ESLint?
